### PR TITLE
HDDS-9807. Consider volume committed space when checking if datanode can host new container

### PR DESCRIPTION
**Pull request by vtutrinov**
## What changes were proposed in this pull request?

The algorithm of detection of the datanode availability for a new container was different between the SCM and datanodes in case of EC buckets. The SCM doesn't take into account the committed bytes count and serves the client with an invalid pipeline and then the datanodes from the provided pipeline will fail to create new containers

The PR proposes a common way of computing the datanode availability for a new container allocation

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-9807

## How was this patch tested?
* manually in a way described in the jira ticket,
* new unit test that checks the datanode is not available to allocate a new container in case of increasing committed bytes count

### CODE REVIEW

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/interfaces/StorageLocationReportMXBean.java [hunk 1]
```diff
@@ -33,6 +33,8 @@ public interface StorageLocationReportMXBean {
 
   long getRemaining();
```
Comment (xichen01) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/interfaces/StorageLocationReportMXBean.java:34]: @vtutrinov @adoroszlai 
What is the definition of `remaining`, I noticed that `remaining` is actually Volume `available`. `available=capacity - reserved - used`.

- So does `committed` have to be reported to the SCM separately.
- Or is it added to `remaining`, which is defined as the amount of space on the Volume that is guaranteed to be available for data storage? `available=capacity - reserved - used - commited`.
We use `available` in a lot of places, and it seems like we should take `commited` into `available`.


https://github.com/apache/ozone/blob/9b3d045a9a2d87b67cc8523d166c3a82c4e05fa5/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/MutableVolumeSet.java#L473

Comment (xichen01) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/interfaces/StorageLocationReportMXBean.java]: Yes, it looks like your initial solution is the more appropriate choice.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/SCMCommonPlacementPolicy.java [hunk 2]
```diff
@@ -308,7 +312,9 @@ public static boolean hasEnoughSpace(DatanodeDetails datanodeDetails,
 
     if (dataSizeRequired > 0) {
       for (StorageReportProto reportProto : datanodeInfo.getStorageReports()) {
-        if (reportProto.getRemaining() > dataSizeRequired) {
+        if (hasVolumeEnoughSpace(reportProto.getCapacity(),
+            reportProto.getRemaining(), reportProto.getCommitted(),
+            dataSizeRequired, conf)) {
```
Comment (xichen01) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/SCMCommonPlacementPolicy.java:317]: The configuration used in `getMinVolumeFreeSpace` is the Datanode configuration, what if the configuration is different for different Datanodes or not available in SCM?
Maybe `volumeFreeSpaceToSpare` should also report

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/UsageInfoSubcommand.java [hunk 3]
```diff
@@ -155,6 +155,8 @@ private void printInfo(DatanodeUsage info) {
         + " B", StringUtils.byteDesc(info.getRemaining()));
     System.out.printf("%-13s: %s %n", "Remaining %",
         PERCENT_FORMAT.format(info.getRemainingRatio()));
+    System.out.printf("%-13s: %s (%s) %n", "Committed", info.getCommitted()
```
Comment (xichen01) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/UsageInfoSubcommand.java:158]: `Committed` doesn't seem to be easy to understand outside of the code context, suggesting: `Container Preallocated`. 
And add a `Remaining Allocatable` (remaining - commited)

**Comment by adoroszlai** [comment]
Thanks @vtutrinov for updating the patch.  There is a test failure in `TestSCMCommonPlacementPolicy.testDatanodeIsInvalidInCaseOfIncreasingCommittedBytes`: https://github.com/vtutrinov/ozone/actions/runs/7156214054/job/19485712139#step:5:1750

**Comment by xichen01** [comment]
@vtutrinov The patch looks good.
The only thing we might want to consider at this point is `FreeSpaceToSpare`. Requires the "remaining allocatable space" greater than `FreeSpaceToSpare` when create a new `Container`, so maybe we should add that to the output of the `UsageInfo command` as well, what do you think?

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/UsageInfoSubcommand.java [hunk 4]
```diff
@@ -155,6 +155,14 @@ private void printInfo(DatanodeUsage info) {
         + " B", StringUtils.byteDesc(info.getRemaining()));
     System.out.printf("%-13s: %s %n", "Remaining %",
         PERCENT_FORMAT.format(info.getRemainingRatio()));
+    System.out.printf("%-13s: %s (%s) %n", "Container Pre-allocated",
+        info.getCommitted() + " B", StringUtils.byteDesc(info.getCommitted()));
+    System.out.printf("%-13s: %s (%s) %n", "Remaining Allocatable",
+        (info.getRemaining() - info.getCommitted()) + " B",
+        StringUtils.byteDesc((info.getRemaining() - info.getCommitted())));
+    System.out.printf("%-13s: %s (%s) %n", "Free Space To Spare",
```
Comment (xichen01) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/UsageInfoSubcommand.java:163]: We should make this newly added lines aligned
Such as:
```bash
//..
Remaining    : 4751560704 B (4.43 GB)
Remaining %  : 45.02%
Container Pre-allocated : 0 B (0 B)
Remaining Allocatable   : 4751560704 B (4.43 GB)
Free Space To Spare     : 0 B (0 B)
```

**Comment by xichen01** [comment]
@vtutrinov Thank for updating the patch. LGTM + 1 .

**Comment by adoroszlai** [comment]
@aswinshakil @sadanand48 please review

**Comment by adoroszlai** [comment]
Thanks @vtutrinov for the patch, @sumitagrawl, @xichen01 for the review.

(Since release branch for 1.4 has been cut, we can fix any post-commit comments later, without holding back the upcoming release.)

**Comment by symious** [comment]
@adoroszlai Could you help to confirm if this PR needs to be cherry-picked to ozone-1.4?

**Comment by adoroszlai** [comment]
@symious @myskov I suggest not including it in 1.4.0, as this had little time to be tested.  We can work on a patch realease 1.4.1 soon afterwards.

**Comment by symious** [comment]
@adoroszlai Thank you for the confirmation.

**Comment by szetszwo** [comment]
This may cause NPE; see HDDS-10027.

**Pull request merged**